pkg/adapter/config/encryptionAdapter.go

```diff
@@ -2,29 +2,155 @@ package config
+		}
+		encryptedEntries[configKey] = common.EncryptedDoguConfigValue(encryptedValue)
+	}
+	return encryptedEntries, errors.Join(encryptionErrors...)
+}
+
+func loadPubKeysFromMap(registry etcdRegistry, entries map[common.SensitiveDoguConfigKey]common.SensitiveDoguConfigValue) (map[string]*keys.PublicKey, error) {
+	pubkeys := map[string]*keys.PublicKey{}
+	var getPubKeyErrors []error
+	for configKey, _ := range entries {
+		doguname := string(configKey.DoguName)
+		_, pubkeyKnown := pubkeys[doguname]
+		if !pubkeyKnown {
+			pubkeyFromRegistry, err := resource.GetPublicKey(registry, doguname)
+			if err != nil {
+				getPubKeyErrors = append(getPubKeyErrors, domainservice.NewNotFoundError(err, "could not get public key for dogu %v", doguname))
+				continue
+			} else {
+				pubkeys[doguname] = pubkeyFromRegistry
+			}
+		}
+	}
+	return pubkeys, errors.Join(getPubKeyErrors...)
+}
+
+func (p PublicKeyConfigEncryptionAdapter) Decrypt(
+	ctx context.Context,
+	name common.SimpleDoguName,
+	encryptedValue common.EncryptedDoguConfigValue,
+) (common.SensitiveDoguConfigValue, error) {
+	privateKeySecret, err := p.secrets.Get(ctx, string(name)+"-private", metav1.GetOptions{})
+	if err != nil {
+		return "", domainservice.NewNotFoundError(err, "could not get private key")
+	}
+	privateKey := privateKeySecret.Data["private.pem"]
+	keyPair, err := getKeyPairFromPrivateKey(privateKey, p.registry)
+	if err != nil {
+		return "", domainservice.NewInternalError(err, "could not get key pair for dogu %v", string(name))
+	}
+	decryptedValue, err := keyPair.Private().Decrypt(string(encryptedValue))
+	if err != nil {
+		return "", domainservice.NewInternalError(err, "could not decrypt encrypted value")
+	}
+	return common.SensitiveDoguConfigValue(decryptedValue), nil
+}
+
+func getKeyPairFromPrivateKey(privateKey []byte, registry etcdRegistry) (*keys.KeyPair, error) {
+	keyProvider, err := resource.GetKeyProvider(registry)
+	if err != nil {
+		return &keys.KeyPair{}, domainservice.NewNotFoundError(err, "could not get key provider")
+	}
+	keyPair, err := keyProvider.FromPrivateKey(privateKey)
+	if err != nil {
+		return &keys.KeyPair{}, domainservice.NewInternalError(err, "could not get key pair")
+	}
+	return keyPair, nil
+}
+
+func (p PublicKeyConfigEncryptionAdapter) DecryptAll(ctx context.Context, entries map[common.SensitiveDoguConfigKey]common.EncryptedDoguConfigValue) (map[common.SensitiveDoguConfigKey]common.SensitiveDoguConfigValue, error) {
+	decryptedEntries := map[common.SensitiveDoguConfigKey]common.SensitiveDoguConfigValue{}
+	var decryptionErrors []error
+	keypairs, err := getKeyPairsFromMap(ctx, entries, p.secrets, p.registry)
+	if err != nil {
+		return decryptedEntries, domainservice.NewNotFoundError(err, "could not get key pair(s)")
+	}
+	for configKey, configValue := range entries {
+		doguname := string(configKey.DoguName)
+		keyPair, _ := keypairs[doguname]
```

Comment:
🚫 **[checkstyle]** <gosimple> <sub>reported by [reviewdog](https://github.com/reviewdog/reviewdog) :dog:</sub><br>S1005: unnecessary assignment to the blank identifier